src/solvers/iterators.jl

```diff
@@ -0,0 +1,104 @@
+#
+# SweepIterator
+#
+
+mutable struct SweepIterator
+  sweep_kws
+  region_iter
+  which_sweep::Int
+end
+
+problem(S::SweepIterator) = problem(S.region_iter)
+
+Base.length(S::SweepIterator) = length(S.sweep_kws)
+
+function Base.iterate(S::SweepIterator, which=nothing)
+  if isnothing(which)
+    sweep_kws_state = iterate(S.sweep_kws)
+  else
+    sweep_kws_state = iterate(S.sweep_kws, which)
+  end
+  isnothing(sweep_kws_state) && return nothing
+  current_sweep_kws, next = sweep_kws_state
+
+  if !isnothing(which)
+    S.region_iter = region_iterator(
+      problem(S.region_iter); sweep=S.which_sweep, current_sweep_kws...
+    )
+  end
+  S.which_sweep += 1
+  return S.region_iter, next
+end
+
+function sweep_iterator(problem, sweep_kws)
+  region_iter = region_iterator(problem; sweep=1, first(sweep_kws)...)
+  return SweepIterator(sweep_kws, region_iter, 1)
+end
+
+function sweep_iterator(problem, nsweeps::Integer; sweep_kws...)
+  return sweep_iterator(problem, Iterators.repeated(sweep_kws, nsweeps))
+end
+
+#
+# RegionIterator
+#
+
+@kwdef mutable struct RegionIterator{Problem,RegionPlan}
+  problem::Problem
+  region_plan::RegionPlan
+  which_region::Int = 1
+end
+
+problem(R::RegionIterator) = R.problem
+current_region_plan(R::RegionIterator) = R.region_plan[R.which_region]
+current_region(R::RegionIterator) = current_region_plan(R)[1]
```

Comment:
What's the data structure for the output of `current_region_plan(R)`? Something doesn't feel write to me that the region is accessed with indexing by `1`, maybe it should be a `NamedTuple` and the region could be accessed as `current_region_plan(R).region` or a struct where it is accessed with a function `get_region(current_region_plan(R))`?